starlette/requests.py

```diff
@@ -205,10 +205,33 @@ def method(self) -> str:
     def receive(self) -> Receive:
         return self._receive
 
+    @property
+    def _stream(self) -> typing.List[bytes]:
+        try:
+            return self.scope["extensions"]["starlette"]["stream"]
+        except KeyError:
+            raise AttributeError("_stream")
+
+    @_stream.setter
+    def _stream(self, chunks: typing.List[bytes]) -> None:
+        if "starlette" not in self.scope["extensions"]:
+            self.scope["extensions"]["starlette"] = {}
+        self.scope["extensions"]["starlette"]["stream"] = chunks
+
+    async def _cache_stream(self) -> typing.AsyncGenerator[bytes, None]:
+        _stream: typing.List[bytes] = []
+
+        async for chunk in self.stream():
+            _stream.append(chunk)
+            yield chunk
+
+        if not hasattr(self, "_stream"):
+            self._stream = _stream
+
     async def stream(self) -> typing.AsyncGenerator[bytes, None]:
-        if hasattr(self, "_body"):
-            yield self._body
-            yield b""
+        if hasattr(self, "_stream"):
+            for chunk in self._stream:
+                yield chunk
             return
```

Comment:
It took me a sec to process what is calling what here. I wonder if we could instead do something like:


```python
class CachedStream:
    def __init__(self, receive: Receive) -> None:
         self.chunks = list()
    
    async def __aiter__(self) -> AsyncIterator[bytes]:
        # read the stream and save the chunks
        ...

    @staticmethod
     def from_asgi(self, scope: Scope, receive: Receive) -> BodyStream:
         # if stream exists in scope, retrieve it from there, if not create an store
         ....
```

And then use that in `Request`?

Not sure if that's better, just an idea 😃

Reply:
ya, i can't really speak to what is better or not. This PR has attempted to follow the existing coding patterns and make as small a change as possible. i think any kind of clean up like this should be part of a separate PR to make stream caching more general purpose and a first class citizen, which is a larger scope than this PR and my commitment to this project.